### PR TITLE
Add dimension names to ArrayLikeSpec

### PR DESCRIPTION
**Pull request by timothymillar**
This is a proof of concept for #904.

It adds the option of named dimensions to `ArrayLikeSpec`. Individual dims can be optional, have multiple valid names, or be unnamed using a wildcard ("*"). 

If `dims` is specified without `ndim` then `ndim` will be calculated from `dims`. If they are both specified then they are checked for congruence. Dimension names are checked as part of variable validation and a `DimensionWarning` is given if the observed `dims` don't match the spec. A `DimensionWarning`  is also given if a spec is so vague that there is more than one possible mapping of array dims to spec dims (this is only possible with multiple optional dims and isn't an issue for any current variables).

I've possibly been a bit heavy handed with adding named dimensions to variables and some of these could be rolled back. I also identified a few slightly inconsistent variables that may be worth altering (possible breaking changes):

- `genotype_count` could have "genotypes" as second dimension but we should consider resolving  #911 first.
- `ld_prune_index_to_drop` has a unique dimension "variant_indexes" and holds integer indices for a subset of variants,
  I wonder if this could be replaced with a length "variants" array of booleans to avoid adding a new dimension?
- `variant_linreg_effect` doesn't appear to be used, can it be removed?

It would also be nice to automatically add the `ndim` and `dims` field to the `__doc__` string, is there a more elegant way to do this than just appending within the `__post_init__` method?

### CODE REVIEW

**Comment by timothymillar** [comment]
CI is failing on coverage due to the change made to `test_regenie_loco_regression`. 
@tomwhite I don't quite grok why there is a special case handled [here](https://github.com/pystatgen/sgkit/blob/main/sgkit/stats/association.py#L351) for a covariates array with an outer dimension of length 0? The documentation for this variable states "variables of shape (samples[, covariates])" which makes me think this should check for `covariates.shape[-1]`?

**Comment by tomwhite** [comment]
> CI is failing on coverage due to the change made to `test_regenie_loco_regression`.
> @tomwhite I don't quite grok why there is a special case handled [here](https://github.com/pystatgen/sgkit/blob/main/sgkit/stats/association.py#L351) for a covariates array with an outer dimension of length 0? The documentation for this variable states "variables of shape (samples[, covariates])" which makes me think this should check for `covariates.shape[-1]`?

It looks like this was #509. Does the unit test introduced there still pass with your suggested change?

**Comment by tomwhite** [comment]
> I also identified a few slightly inconsistent variables that may be worth altering (possible breaking changes):
> 
> * `genotype_count` could have "genotypes" as second dimension but we should consider resolving  [genotype_count_spec is inconsistent with VCF spec #911](https://github.com/pystatgen/sgkit/issues/911) first.

I think this would be a good change. You could do this as a follow on by omitting `dims` in this PR.

> * `ld_prune_index_to_drop` has a unique dimension "variant_indexes" and holds integer indices for a subset of variants,
>   I wonder if this could be replaced with a length "variants" array of booleans to avoid adding a new dimension?

Possibly, although that code path has had some performance work done on it, so changing it might have an unknown impact unless we measure it again. Another option would be not to name the dimension, and open an issue with a comment in the code.

> * `variant_linreg_effect` doesn't appear to be used, can it be removed?

Yes I think so.

File: sgkit/variables.py [hunk 1]
```diff
@@ -671,7 +731,7 @@ def _check_field(
 stat_Garud_h1, stat_Garud_h1_spec = SgkitVariables.register_variable(
     ArrayLikeSpec(
         "stat_Garud_h1",
-        ndim={1, 2},
+        dims=("windows", "cohorts"),
```
Comment (tomwhite) [inline sgkit/variables.py:734]: It looks like the Garud stats are always windowed, so we should change their api doc to reflect that.

**Comment by timothymillar** [comment]
> It looks like this was https://github.com/pystatgen/sgkit/issues/509. Does the unit test introduced there still pass with your suggested change?

That PR seems to relate to  `gwas_linear_regression` where I was referring to the code in `regenie_loco_regression`. The APIs of these methods are inconsistent in terms of the covariates parameter. The former expects a list of covariate variables where the latter expects a single covariates variable. They have very similar checks for `len(covariates) == 0` which actually do quite different things:

- When `gwas_linear_regression` checks `if len(covariates) == 0` it is checking how many "covariates" variables have been specified (i.e., the number of arrays which will be concatenated).
- When `regenie_loco_regression` checks `if len(covariates) == 0` it is checking the length of the first dimension of the single "covariates" array.

The check in `regenie_loco_regression` is odd to me because the description of the "covariates" array states that the first dimension should always have length=samples. Maybe the intent was to match the `gwas_linear_regression` API?

**Comment by timothymillar** [comment]
Thanks @tomwhite, I've avoided the awkward variables for now.

**Review by tomwhite — APPROVED** [review]

**Comment by timothymillar** [comment]
I'm thinking these changes should warrant a second reviewer or a bit more consensus?

**Comment by jeromekelleher** [comment]
I wonder if @eric-czech or @ravwojdyla would like to take a quick look? I think they were heavily involved in the design of the variable naming/validation system.

**Review by ravwojdyla — APPROVED** [review]
This looks great, nice and clean code btw 👍

**Pull request merged**